doc/exceptions.md

```diff
@@ -0,0 +1,365 @@
+rescue NoSuch
+  puts 'Rescued an exception!'
+else
+  puts 'No exception raised.'
+end
+```
+
+Output:
+
+```
+Begin.
+No exception raised.
+```
+
+#### Ensure Clause
+
+The ensure clause:
+
+- Starts with an `ensure` statement.
+- Contains code that is to be executed
+  regardless of whether an exception is raised,
+  and regardless of whether a raised exception is handled.
+- Ends with the first following `end` statement.
+
+```
+def foo(boom: false)
+  puts 'Begin.'
+  raise 'Boom!' if boom
+rescue
+  puts 'Rescued an exception!'
+else
+  puts 'No exception raised.'
+ensure
+  puts 'Always do this.'
+end
+
+foo(boom: true)
+foo(boom: false)
+```
+
+Output:
+
+```
+Begin.
+Rescued an exception!
+Always do this.
+Begin.
+No exception raised.
+Always do this.
+```
+
+#### End Statement
+
+The `end` statement ends the handler.
+
+Code following it is reached only if any raised exception is rescued.
+
+#### Begin-Less \Exception Handlers
+
+As seen above, an exception handler may be implemented with `begin` and `end`.
+
+An exception handler may also be implemented as:
+
+- A method body:
+
+    ```
+    def foo(boom: false) # Serves as beginning of exception handler.
+      puts 'Begin.'
+      raise 'Boom!' if boom
+    rescue
+      puts 'Rescued an exception!'
+    else
+      puts 'No exception raised.'
+    end                  # Serves as end of exception handler.
+    ```
+
+- A block:
+
+    ```
+    Dir.chdir('.') do |dir| # Serves as beginning of exception handler.
+      raise 'Boom!'
+    rescue
+      puts 'Rescued an exception!'
+    end                     # Serves as end of exception handler.
+    ```
+
+#### Re-Raising an \Exception
+
+It can be useful to rescue an exception, but allow its eventual effect;
+for example, a program can rescue an exception, log data about it,
+and then "reinstate" the exception.
+
+This may be done via the `raise` method, but in a special way;
+a rescuing clause:
+
+  - Captures an exception.
+  - Does whatever is needed concerning the exception (such as logging it).
+  - Calls method `raise` with no argument,
+    which raises the rescued exception:
+
+```
+begin
+  1 / 0
+rescue ZeroDivisionError
+  # Do needful things (like logging).
+  raise # Raised exception will be ZeroDivisionError, not RuntimeError.
+end
+```
+
+Output:
+
+```
+ruby t.rb
+t.rb:2:in `/': divided by 0 (ZeroDivisionError)
+        from t.rb:2:in `<main>'
+```
+
+#### Retrying
+
+It can be useful to retry a begin clause;
+for example, if it must access a possibly-volatile resource
+(such as a web page),
+it can be useful to try the access more than once
+(in the hope that it may become available):
+
+```
+require 'open-uri'
+begin
+  retries ||= 0
+  tempfile = URI.open("https://docs.ruby-lang.org/en/master/Exception.html")
+  # Do something with the retrieved data.
+rescue OpenURI::HTTPError
+  retry if (retries += 1) < 5
+else
+  puts "Got it, with #{retries} retries."
+end
+```
+
+```
+Got it, with 0 retries.
```

Comment:
This example shows a real-life example of how to use retries but doesn't really show how retries work because it succeeded on the first try. We should use a more artificial example that demonstrates how retries work.

Reply:
Done.